allennlp/data/data_loaders/multi_process_data_loader.py

```diff
@@ -0,0 +1,304 @@
+                self._instances = []
+
+            if self.num_workers <= 0:
+                # Just read all instances in main process.
+                for instance in Tqdm.tqdm(
+                    self.reader.read(self.data_path), desc="loading instances"
+                ):
+                    self.reader.apply_token_indexers(instance)
+                    if not self.lazy:
+                        self._instances.append(instance)  # type: ignore
+                    if self._vocab is not None:
+                        instance.index_fields(self._vocab)
+                    yield instance
+            else:
+                queue: mp.JoinableQueue = mp.JoinableQueue(self._INSTANCE_QUEUE_SIZE)
+                workers = self._start_instance_workers(queue)
+
+                for instance in Tqdm.tqdm(self._gather_instances(queue), desc="loading instances"):
+                    if not self.lazy:
+                        self._instances.append(instance)  # type: ignore
+                    yield instance
+
+                self._join_workers(workers)
+
+    def index_with(self, vocab: Vocabulary) -> None:
+        self._vocab = vocab
+        if self._instances:
+            for instance in self._instances:
+                instance.index_fields(vocab)
+
+    def _start_instance_workers(self, queue: mp.JoinableQueue) -> List[mp.Process]:
+        workers: List[mp.Process] = []
+        for worker_id in range(self.num_workers):
+            worker = mp.Process(target=self._instance_worker, args=(worker_id, queue), daemon=True)
+            worker.start()
+            workers.append(worker)
+        return workers
+
+    def _join_workers(self, workers: List[mp.Process]) -> None:
+        for worker in workers:
+            # TODO: handle errors if any of the workers crash.
+            worker.join(1)
+            if worker.is_alive():
+                logger.warning("Worker is still alive, killing now")
+                worker.terminate()
+
+    def _gather_instances(self, queue: mp.JoinableQueue) -> Iterable[Instance]:
+        done_count: int = 0
+        while done_count < self.num_workers:
+            instances_chunk: Optional[List[Instance]]
+            for instances_chunk in iter(queue.get, None):
+                yield from instances_chunk
+                queue.task_done()
+            queue.task_done()
+            # Every time we encounter an empty list, thats means a worker has finished.
+            done_count += 1
+
+    def _instance_worker(self, worker_id: int, queue: mp.JoinableQueue) -> None:
+        self.reader._set_worker_info(WorkerInfo(self.num_workers, worker_id))
+
+        def index_fields(instance: Instance) -> Instance:
+            self.reader.apply_token_indexers(instance)
+            if self._vocab is not None:
+                instance.index_fields(self._vocab)
+            return instance
+
+        instances = (index_fields(instance) for instance in self.reader.read(self.data_path))
```

Comment:
This needs error handling, in case `read()` throws something. If you want to peek at how I did it in bettermap, the code is here: https://github.com/allenai/bettermap/blob/master/bettermap.py#L84

Reply:
This is what happens on the receiving end: https://github.com/allenai/bettermap/blob/master/bettermap.py#L107

Reply:
Ah, and then there is this bit: https://github.com/allenai/bettermap/blob/master/bettermap.py#L132

That makes sure that all the other workers get killed when something goes wrong in one of them.

Reply:
Thanks, I'll check that out